### PR TITLE
[NET-635][system-probe] Report count of different types of DNS errors

### PR DESCRIPTION
**Pull request by mbakht**
### What does this PR do?
Payload PR: https://github.com/DataDog/agent-payload/pull/63
JIRA Ticket: https://datadoghq.atlassian.net/browse/NET-635

## Problem 
Currently, when we receive a `DNS` response, we categorize them into two groups based on the `rcode`:
- If `rcode == 0`, we increment the count for `SuccessfulResponses`
- Otherwise, we increment the count `FailedResponses`

It turns out that this information is too coarse grained. For a failure, the user wants to know the specific type of the error.

## Solution
In this PR, we update the code to keep track of the `count` of responses received with different `rcode`s. 

|  rcode | count   |
|---|---|
|  0 | 1   |
|  3 | 5  |
|  ... | ...  |

Instead of keeping separate count for `Successes` and `Failures`, we use only this map in the `Snooper`. However, when sending data to the backend, we still need to populate the fields `DNSSuccessfulResponses` and `DNSFailedResponses` for backward compatibility with the backend. We will gradually phase them out.

## Additional changes

- `github.com/miekg/dns` has been upgraded to `v1.1.31`. This is because during testing, we want to reuse the same TCP connection for multiple queries with the function `ExchangeWithConn`. This is not available in the older version we were using.
- There is no way to force`SERVFAIL` response from a regular DNS server. So, in the test for failures, we now spin up a local DNS server that always return `SERVFAIL` irrespective of query. 

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

- Start the `system-probe`. (`sudo ./bin/system-probe/system-probe` on the Agent VM).
- Do a `curl` for a legit domain. ` curl cnn.com`
- Do `curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | jq .`

There should be some output of the following form:

```
    {
      "pid": 644,
      "laddr": {
        "ip": "10.0.2.15",
        "port": 59352,
        "containerId": "",
        "hostId": "0"
      },
      "raddr": {
        "ip": "4.2.2.2",
        "port": 53,
        "containerId": "",
        "hostId": "0"
      },
      "family": "v4",
      "type": "udp",
      "pidCreateTime": "0",
      "lastBytesSent": "59",
      "lastBytesReceived": "200",
      "lastRetransmits": 0,
      "direction": "outgoing",
      "netNS": 4026531992,
      "remoteNetworkId": "",
      "ipTranslation": null,
      "rtt": 0,
      "rttVar": 0,
      "intraHost": false,
      "dnsSuccessfulResponses": 1,
      "dnsFailedResponses": 0,
      "dnsTimeouts": 0,
      "dnsSuccessLatencySum": "66692",
      "dnsFailureLatencySum": "0",
      "dnsCountByRcode": {
        "0": 1
      },
      "lastTcpEstablished": 0,
      "lastTcpClosed": 0
    }
```

Check `dnsCountByRcode`. It should have `1` for key `0` and it should match `dnsSuccessfulResponses`. 
Similarly, we can do curl for an invalid domain (e.g `curl asdfsfs`). The output should be something like the following:

```
"dnsCountByRcode": {
        "3": 1
      },
```

### CODE REVIEW

**Review by heyronhay — CHANGES_REQUESTED** [review]
Mind filling in the PR description text - it's nice to have before the PR is put up, especially as we get new people (like myself :) ) that don't have the context already that we want to review code to help get that context.  

Otherwise looks good, although it would be nice to have another test for a different type of error to exercise the code a bit better.

**Comment by mbakht** [comment]
> Mind filling in the PR description text - it's nice to have before the PR is put up, especially as we get new people (like myself :) ) that don't have the context already that we want to review code to help get that context.

Thanks for the suggestion. I have fleshed out the description section. It is not only for new people - I find it difficult to figure out why I did something in my __own__ PR if I look at it few months later. So it is always good to have it and I need to get into the habit of writing descriptions for all my PRs. Please continue to ping me whenever you don't see one in any of my PRs.

> Otherwise looks good, although it would be nice to have another test for a different type of error to exercise the code a bit better.

Added a test to generate `SERVFAIL` errors though it turned out to be much more difficult than I anticipated. The queries can't be tweaked to force this kind of response. So I ended up using a dumb local server.

File: pkg/network/dns_stats.go [hunk 1]
```diff
@@ -8,12 +8,10 @@ import (
...
-	// More stats like latency, error, etc. will be added here later
-	successfulResponses uint32
-	failedResponses     uint32
-	successLatencySum   uint64 // Stored in µs
-	failureLatencySum   uint64
-	timeouts            uint32
+	successLatencySum uint64 // Stored in µs
+	failureLatencySum uint64
+	timeouts          uint32
+	countByRcode      map[uint8]uint32
```
Comment (leeavital) [inline pkg/network/dns_stats.go:14]: Why is the rcode   uint16 in the protobuf format and a uint8 here?

File: pkg/network/state_test.go [hunk 2]
```diff
@@ -1145,8 +1147,14 @@ func TestDNSStatsWithMultipleClients(t *testing.T) {
...
-	stats := make(map[dnsKey]dnsStats)
-	stats[dKey] = dnsStats{successfulResponses: 1}
+
+	getStats := func() map[dnsKey]dnsStats {
+		stats := make(map[dnsKey]dnsStats)
+		countByRcode := make(map[uint8]uint32)
+		countByRcode[uint8(layers.DNSResponseCodeNoErr)] = 1
+		stats[dKey] = dnsStats{countByRcode: countByRcode}
+		return stats
+	}
```
Comment (leeavital) [inline pkg/network/state_test.go:1157]: Why do we need to keep copying `stats` now?

File: pkg/network/dns_stats.go (hunk 1 above)
Comment (leeavital) [reply pkg/network/dns_stats.go:14]: Sorry, this comment was addressed in agent-payload. Please ignore.

File: pkg/network/state_test.go (hunk 2 above)
Comment (mbakht) [reply pkg/network/state_test.go:1157]: `stats` now has a `map` as its member. So if we reuse the same `stats` object, the map would get modified across different calls to `state.Connections`.

**Review by hmahmood — APPROVED** [review]

**Comment by mbakht** [comment]
@heyronhay , can you please have a look at this PR again? This is blocked on your approval.

File: pkg/network/dns_snooper_test.go [hunk 3]
```diff
@@ -166,28 +191,46 @@ func sendDNSQuery(
...
+		conn, err := dnsClient.Dial(dnsHost)
+		require.NoError(t, err)
+		for _, domain := range domains {
+			msg.SetQuestion(mdns.Fqdn(domain), mdns.TypeA)
+			rep, _, _ := dnsClient.ExchangeWithConn(msg, conn)
+			reps = append(reps, rep)
+		}
+	} else { // UDP
+		for _, domain := range domains {
+			// msg.SetQuestion(mdns.Fqdn(domains[0]), mdns.TypeA)
```
Comment (heyronhay) [inline pkg/network/dns_snooper_test.go:207]: Is this to be deleted, or just commented out while you change something?

**Review by heyronhay — APPROVED** [review]
Shallow review since I lack a lot (all!) of the context, but looks solid to me

File: pkg/network/dns_snooper_test.go (hunk 3 above)
Comment (mbakht) [reply pkg/network/dns_snooper_test.go:207]: Good catch. Will delete it.

**Pull request merged**